numpy/core/src/private/ufunc_override.c

```diff
@@ -0,0 +1,532 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define NO_IMPORT_ARRAY
+
+#include "npy_pycompat.h"
+#include "numpy/ufuncobject.h"
+#include "get_attr_string.h"
+#include "npy_import.h"
+
+#include "ufunc_override.h"
+
+static void
+normalize___call___args(PyUFuncObject *ufunc, PyObject *args,
+                    PyObject **normal_args, PyObject **normal_kwds,
+                    int nin)
+{
+    /* ufunc.__call__(*args, **kwds) */
+    int i;
+    int not_all_none;
+    int nargs = PyTuple_GET_SIZE(args);
+    PyObject *obj = PyDict_GetItemString(*normal_kwds, "sig");
+
+    /* ufuncs accept 'sig' or 'signature' normalize to 'signature' */
+    if (obj != NULL) {
+        Py_INCREF(obj);
+        PyDict_SetItemString(*normal_kwds, "signature", obj);
+        PyDict_DelItemString(*normal_kwds, "sig");
+    }
+
+    *normal_args = PyTuple_GetSlice(args, 0, nin);
+
+    /* If we have more args than nin, they must be the output variables.*/
+    if (nargs > nin) {
+        for (i=nin; i < nargs; i++) {
+            not_all_none = (PyTuple_GET_ITEM(args, i) != Py_None);
+            if (not_all_none) {
+                break;
+            }
+        }
+        if (not_all_none) {
+            obj = PyTuple_GetSlice(args, nin, nargs);
+            PyDict_SetItemString(*normal_kwds, "out", obj);
+            Py_DECREF(obj);
+        }
```

Comment:
As mentioned elsewhere, I'd be in favor of just always doing this (not conditioning on `not_all_none`), since it makes implementing things correctly easier.

Of course, there'd then need to be a `if (kwds['out'] == None) kwds['out'] = (None,)* nout` case.